qemu/tests/numa_cpu.py

```diff
@@ -0,0 +1,132 @@
+        """
+        numa_info_guest = NumaInfo(session=session)
+        nodes_guest = numa_info_guest.online_nodes
+        numa_cpu_guest = []
+        for node in nodes_guest:
+            numa_cpus = numa_info_guest.online_nodes_cpus[node]
+            numa_cpus = set([int(v) for v in numa_cpus.split()])
+            numa_cpu_guest.append(numa_cpus)
+        return numa_cpu_guest
+
+    def get_expected_numa_cpus():
+        """
+        Get the expected numa cpus for every node according to the qemu cli
+        """
+
+        def _get_boundary(value, max_value, weight):
+            min_boundary = int(value if value else 0) * weight
+            max_boundary = int(value if value else (max_value - 1)) * weight
+            return (min_boundary, max_boundary)
+
+        cpus = params.objects("guest_numa_cpus")
+        expected_cpus = []
+        tmp = {}
+        for cpu in cpus:
+            nodeid = params["numa_cpu_nodeid_%s" % cpu]
+            socket = params.get("numa_cpu_socketid_%s" % cpu)
+            die = params.get("numa_cpu_dieid_%s" % cpu)
+            core = params.get("numa_cpu_coreid_%s" % cpu)
+            thread = params.get("numa_cpu_threadid_%s" % cpu)
+            if vm_arch in ('x86_64', 'i386'):
+                socket_min, socket_max = _get_boundary(socket, vcpu_sockets, socket_weight)
+                die_min, die_max = _get_boundary(die, vcpu_dies, die_weight)
+                core_min, core_max = _get_boundary(core, vcpu_cores, core_weight)
+                thread_min, thread_max = _get_boundary(thread, vcpu_threads, thread_weight)
+                cpu_min = socket_min + die_min + core_min + thread_min
+                cpu_max = socket_max + die_max + core_max + thread_max
+                if nodeid in tmp.keys():
+                    tmp[nodeid] += range(cpu_min, cpu_max+1, 1)
+                else:
+                    tmp[nodeid] = range(cpu_min, cpu_max+1, 1)
+            elif vm_arch in ('ppc64le', 'ppc64'):
+                if nodeid in tmp.keys():
+                    tmp[nodeid] += [int(core)]
+                else:
+                    tmp[nodeid] = [int(core)]
+        for item in sorted(tmp.items(), key=lambda item:item[0]):
+            expected_cpus.append(set(sorted(item[1])))
+        return expected_cpus
+
+    def get_hotpluggable_cpus():
+        """
+        Parse output of "query-hotpluggable-cpus", get the cpus that bind to
+        special node.
+        """
+        hotpluggable_cpus = []
+        tmp = {}
+        out = vm.monitor.info("hotpluggable-cpus")
+        for vcpu_info in out:
+            vcpu_info = vcpu_info["props"]
+            nodeid = vcpu_info.get("node-id")
+            if nodeid is not None:
+                socketid = vcpu_info["socket-id"]
+                dieid = vcpu_info["die-id"]
+                coreid = vcpu_info["core-id"]
+                threadid = vcpu_info["thread-id"]
+                cpu_index = socketid * socket_weight + dieid * die_weight \
```

Comment:
I think you can get cpu-index from `query-cpus-fast`
```json
{"execute": "query-cpus-fast"}
{
    "return": [
        {
            "arch": "ppc",
            "thread-id": 40819,
            "props": {
                "core-id": 0,
                "node-id": 0
            },
            "qom-path": "/machine/unattached/device[1]/thread[0]",
            "cpu-index": 0,
            "target": "ppc64"
        },
        {
            "arch": "ppc",
            "thread-id": 40820,
            "props": {
                "core-id": 1,
                "node-id": 0
            },
            "qom-path": "/machine/unattached/device[2]/thread[0]",
            "cpu-index": 1,
            "target": "ppc64"
        },
        {
            "arch": "ppc",
            "thread-id": 40821,
            "props": {
                "core-id": 2,
                "node-id": 0
            },
            "qom-path": "/machine/unattached/device[3]/thread[0]",
            "cpu-index": 2,
            "target": "ppc64"
        },
        {
            "arch": "ppc",
            "thread-id": 40822,
            "props": {
                "core-id": 3,
                "node-id": 0
            },
            "qom-path": "/machine/unattached/device[4]/thread[0]",
            "cpu-index": 3,
            "target": "ppc64"
        }
    ]
}
```

Reply:
@PaulYuuu  The purpose of this function is `Check hotpluggable cpus in qmp`, so it check the output of `query-hotpluggable-cpus`. 
Get socket/die/core/thread from `query-hotpluggable-cpus` but cpu index from `query-cpus-fast` is not a good choice. cpu index and socket/die/core/thread are one-to-one, so I think we need keep the current method.

Reply:
> @PaulYuuu The purpose of this function is `Check hotpluggable cpus in qmp`, so it check the output of `query-hotpluggable-cpus`.
> Get socket/die/core/thread from `query-hotpluggable-cpus` but cpu index from `query-cpus-fast` is not a good choice. cpu index and socket/die/core/thread are one-to-one, so I think we need keep the current method.

If so, I think this method is difficult to calculate cpu_index. On ppc, the cpu_index is `core-id * threads`.

Reply:
Yes, this is the issue you mentioned in another comment `Only coreid is available on ppc.`, will update it later.